pkg/network/listener.go

```diff
@@ -223,12 +224,27 @@ func (l *listener) readMsgEventLoop(lctx context.Context) {
 }
 
 // Shutdown stop accepting new connections and graceful close the existing connections
-func (l *listener) Shutdown() error {
-	changed, err := l.stopAccept()
-	if changed {
+func (l *listener) Shutdown(lctx context.Context) error {
+	// #2220: An exception may occur when a new connection in the gracefully shutdown during non-hot upgrade
+	// If current state is upgrading, it means old Mosn exits after new Mosn starts successfully.
+	if stagemanager.GetState() == stagemanager.Upgrading {
+		// If it is upgrade, stop accept new connection only,
+		changed, err := l.stopAccept()
+		if changed {
+			l.cb.OnShutdown()
+		}
+		return err
+	} else {
+		// If it is not upgrade, close the listener first to avoid new connection establishment.
+		// These newly established connections will not be processed by mosn before listener closed, and the requests on the connection will timeout.
+		//
+		// Close listener should not close existsing connections.
+		err := l.Close(lctx)
+
+		// Close the existing connections if it support graceful close.
```

Comment:
If the `bindToPort` of the listener is false, whether it is necessary to call `OnShutdown` and it is safe?

Reply:
Added a bindToPort status check here to keep it consistent with the original logic